### PR TITLE
New Api Functions

### PR DESCRIPTION
**Pull request by gchoqueux**
**Api functions**
- Add `ApiGlobe.setZoomLevel()`
- Add` ApiGlobe.getZoomScale()`
- Add `ApiGlobe.setZoomScale(zoomScale)`
- Add `ApiGlobe.pan(pVector)`
- Partial Fix bug `ApiGlobe.removeImageryLayer()`

**Framework**
- Three js v80 to 82

**Core**
- SSE
 Remove small oblique weight in Screen space error compute [(see page 372 '3D engine design for virtual globes') ](http://www.virtualglobebook.com/3DEngineDesignForVirtualGlobesSection121.pdf)
 Oblique weight is wrong for tiles with big altitude

### CODE REVIEW

**Comment by nosy-b** [comment]
- Error while calling `itowns.viewer.setZoomScale(zoomScale);` directly after loading the globe. If user click and move it then it works.
- `ApiGlobe.prototype.getZoomLevel` has bad documentation and sample code. It doesn"t take any parameters.

**Comment by gchoqueux** [comment]
@nosy-b see new commits
make `npm install` to check `THREE v82`

**Comment by nosy-b** [comment]
`ApiGlobe.SetCenter()` works but it doesn't keep camera upVector to current direction (North by default)

Other functions seems to work fine. Remember to add documentation for all functions (at least API ones)

**Comment by gchoqueux** [comment]
> `ApiGlobe.SetCenter()` works but it doesn't keep camera upVector to current direction (North by default)
> use ApiGlobe.setCenterAdvanced() for the moment.
> Add a issue for this feature

The branch is updated

File: package.json [hunk 1]
```diff
@@ -28,7 +28,7 @@
     "paralleljs": "^0.2.1",
     "simd": "^2.0.0",
     "string_format": "^0.0.6",
-    "three": "^0.80.0",
+    "three": "^0.82.0",
```
Comment (peppsac) [inline package.json]: Is the change intentionnal?
I don't mind upgrading three.js, but you should write about this change in the commit message

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 2]
```diff
@@ -591,13 +607,111 @@ ApiGlobe.prototype.setRange = function(pRange /*, bool*/ ) {
...
+};
+
+
+/**
+ * Return the current zoom scale at the central point of the view.
+ *
+ * @param      {number}  pitch   Screen pitch, in millimeters ; 0.28 by default
+ * @return     {number}  The zoom scale.
+ */
+ApiGlobe.prototype.getZoomScale = function(pitch) {
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:647]: What is the difference betweem zoom _level_ and zoom _scale_ ?

File: src/Scene/LayersConfiguration.js [hunk 3]
```diff
@@ -44,76 +44,76 @@ function defaultState(seq) {
...
 LayersConfiguration.prototype.addGeometryLayer = function(layer) {
     this.geometryLayers.push(layer);
     this.layersState[layer.id] = defaultState();
-}
+};
 
 LayersConfiguration.prototype.removeColorLayer = function(id) {
     if (this.layersState[id]) {
-        this.colorLayers.filter(function(l) {
+        this.colorLayers = this.colorLayers.filter(function(l) {
```
Comment (peppsac) [inline src/Scene/LayersConfiguration.js:61]: The changes to this file should be done in a separate commit.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 2 above)
Comment (gchoqueux) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:647]: ZoomLevel, zoom camera based on level (tile matrix)
ZoomScale, zoom camera based on scale's map

**Comment by gchoqueux** [comment]
branch updated

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 2 above)
Comment (peppsac) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:647]: So zoom level is based on WMTS' notion of levels right?

Comment (gchoqueux) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:647]: Zoom level is based level's tile mesh

**Comment by peppsac** [comment]
It's hard to test this PR: the API is not accessible using the GUI.
Maybe you could add a new `dat.gui` window with a preconfigured action for each API function?

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 2 above)
Comment (peppsac) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:647]: I get the same result when using `itowns.viewer.setZoomScale(12)` or `itowns.viewer.setZoomScale(0.0000001)`.
But if I use `itowns.viewer.setZoomScale(0.00000001)` I get a black screen and then `setZoomScale` is broken (`rayon` is always `NaN`).

Comment (gchoqueux) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:647]: branch updated

**Comment by gchoqueux** [comment]
> Maybe you could add a new dat.gui window with a preconfigured action for each API function?

Yes, in new PR.
If not, I use console's navigator to test API

**Comment by gchoqueux** [comment]
branch updated and rebased

**Comment by gchoqueux** [comment]
r?

**Comment by peppsac** [comment]
Maybe we could start using label to mark PR that are ready for reviews?

**Comment by peppsac** [comment]
Anyway I started looking at this one but I got lost investigating errors in `getPickingPosition` while testing:

``` js
window.itowns.viewer.setZoomLevel(12)
GlobeControls.js:595 Uncaught TypeError: Cannot read property 'distanceTo' of undefined(…)
```

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 4]
```diff
@@ -580,13 +594,117 @@ ApiGlobe.prototype.setRange = function(pRange /*, bool*/ ) {
...
+/**
+ * Gets the current zoom level, which is an index in the logical scales predefined for the application.
+ * The higher the level, the closer to the ground.
+ * The level is always in the [getMinZoomLevel(), getMaxZoomLevel()] range.
+ *
+ * @param      {number}  zoom    The zoom
  */
+ApiGlobe.prototype.setZoomLevel = function(zoom) {
+
+    zoom = Math.max(2,zoom);
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Why do you use a hard-coded level?
You should compute `zoom` using `[getMinZoomLevel(), getMaxZoomLevel()]`

(and I just tested this: these 2 methods are broken :-( )

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 5]
```diff
@@ -456,16 +469,17 @@ ApiGlobe.prototype.getHeading = function() {
  */
 
 ApiGlobe.prototype.getRange = function() {
+    return this.scene.currentControls().getRange();
+};
+
+ApiGlobe.prototype.getRangeFromEllispoid = function() {
 
+    // TODO: error is distance is big with ellipsoid.intersection(ray) because d < 0
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:484]: Typo in the method name `getRangeFromEllispoid` → `getRangeFromEllipsoid`
Also : no documentation? what is the difference between getRange and `getRangeFromEllipsoid`? why/when should I use one or the other?

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 6]
```diff
@@ -580,13 +594,117 @@ ApiGlobe.prototype.setRange = function(pRange /*, bool*/ ) {
...
+    pitch = (pitch || 0.28)/1000;
+
+    let camera = this.scene.currentCamera();
+    let center = this.scene.currentControls().getCenter();
+    let rayon = center.length();
+    let range = center.distanceTo(camera.camera3D.position);
+    let distance = rayon + range;
+    let alpha = camera.FOV / 180 * Math.PI * 0.5;
+    let phi = Math.PI - Math.asin(distance / rayon * Math.sin(alpha));
+    let projection;
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:673]: Would you mind adding some comments about what is going on here?
It's not obvious, even if you're using sensible variable names.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 7]
```diff
@@ -580,13 +594,117 @@ ApiGlobe.prototype.setRange = function(pRange /*, bool*/ ) {
...
+        projection = distance * 2 * Math.tan(alpha);
+
+    } else {
+
+        /* // develop operation
+        var beta = Math.PI - ( phi + alpha);
+        projection = rayon * Math.sin(beta) * 2.0;
+        // factorisation -->> */
+        projection = 2.0 * rayon * Math.sin(phi + alpha);
+
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Unneeded empty lines + dead code

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 8]
```diff
@@ -580,13 +594,117 @@ ApiGlobe.prototype.setRange = function(pRange /*, bool*/ ) {
...
+    // Screen pitch, in millimeters
+    pitch = (pitch || 0.28)/1000;
+
+    let camera = this.scene.currentCamera();
+    let projection = camera.height * pitch / zoomScale;
+    let rayon = this.scene.currentControls().getCenter().length();
+    let alpha = camera.FOV / 180 * Math.PI * 0.5;
+    let distance;
+    let sinBeta = projection / (2 * rayon);
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:729]: Same: this really needs some comments.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 9]
```diff
@@ -580,13 +594,117 @@ ApiGlobe.prototype.setRange = function(pRange /*, bool*/ ) {
...
 /**
+ * { Displaces the central point to a specific amount of pixels from its current position.
+ *  The view flies to the desired coordinate, i.e.is not teleported instantly. Note : The results can be strange in some cases, if ever possible, when e.g.the camera looks horizontally or if the displaced center would not pick the ground once displaced.}
+ *
+ * @param      {<type>}  pVector  The vector
+ */
+ApiGlobe.prototype.pan = function(pVector) {
+
+    this.scene.currentControls().pan(pVector.x,pVector.y);
+
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Remove empty lines

File: src/Globe/Globe.js [hunk 10]
```diff
@@ -219,10 +220,18 @@ Globe.prototype.getZoomLevel = function( /*id*/ ) {
...
 
     }();
+
     this.tiles.children[0].traverseVisible(cO);
     return cO();
 };
 
+
+Globe.prototype.computeDistanceForZoomLevel = function(zoom,camera) {
+
```
Comment (peppsac) [inline src/Globe/Globe.js]: 3 hard-coded values in a single line: 2 (min level), 18 (max level), 6.0 (sse).

Proposal  : 
- `2` becomes `this.tiles.minLevel`
- `18` becomes `this.tiles.maxLevel`
- `6.0` : you could add a `export const SSE_SUBDIVISION_THRESHOLD` in `NodeProcess.js` and replace all usage of `6.0` in itowns: `NodeProcess.js`, `QuadTree.js` and here.

(and empty blank lines)

File: src/Renderer/Camera.js [hunk 11]
```diff
@@ -124,11 +124,13 @@ Camera.prototype.computeNodeSSE = function(node) {
     var boundingSphere = node.geometry.boundingSphere;
     var distance = Math.max(0.0, (this.camera3D.position.distanceTo(node.centerSphere) - boundingSphere.radius));
     // Added small oblique weight (distance is not enough, tile orientation is needed)
+    /*
```
Comment (peppsac) [inline src/Renderer/Camera.js:128]: You're switching back to the old SSE?
Documenting why (in commit message) would be interesting - and maybe change the commit's title btw: you're doing more than juste `AddNew Api Functions`.

File: src/Renderer/ThreeExtented/GlobeControls.js [hunk 12]
```diff
@@ -471,41 +469,39 @@ function GlobeControls(camera, domElement, engine) {
...
-            // transform to local globeTarget
-            offset.applyMatrix4( globeTarget.matrixWorldInverse );
-        }
-
-        offGT.copy(globeTarget.position);
 
-        if (state === CONTROL_STATE.MOVE_GLOBE) {
+            movingGlobeTarget.copy(this.getCenter()).applyQuaternion(quatGlobe);
+            this.camera.position.copy(snapShotCamera.position).applyQuaternion(quatGlobe);
+            this.camera.up.copy(movingGlobeTarget.clone().normalize());
```
Comment (peppsac) [inline src/Renderer/ThreeExtented/GlobeControls.js:478]: Again: adding a single line of comment explaining what you're doing here would be useful.
Examples: 

``` js
// Rotate globe then update camera to keep camera's center aligned
// Pan camera along globe's normal
```

File: src/Renderer/ThreeExtented/GlobeControls.js [hunk 13]
```diff
@@ -471,41 +469,39 @@ function GlobeControls(camera, domElement, engine) {
...
+
+        } else if (state === CONTROL_STATE.PANORAMIC) {
+
+            this.camera.worldToLocal(movingGlobeTarget);
+            var normal = this.camera.position.clone().normalize().applyQuaternion(this.camera.quaternion.clone().inverse());
+            quaterPano.setFromAxisAngle(normal, sphericalDelta.theta).multiply(quaterAxis.setFromAxisAngle(axisX, sphericalDelta.phi));
+            movingGlobeTarget.applyQuaternion(quaterPano);
+            this.camera.localToWorld(movingGlobeTarget);
+            this.camera.up.copy(movingGlobeTarget.clone().normalize());
+
```
Comment (peppsac) [inline src/Renderer/ThreeExtented/GlobeControls.js]: Same : add comment please

File: src/Renderer/ThreeExtented/GlobeControls.js (hunk 12 above)
Comment (peppsac) [reply src/Renderer/ThreeExtented/GlobeControls.js:478]: The controls are still buggy (even after this PR: try clicking outside the globe), so adding comments explaining what the code is supposed to do will help us fix the problems.

File: src/Renderer/ThreeExtented/GlobeControls.js [hunk 14]
```diff
@@ -471,41 +469,39 @@ function GlobeControls(camera, domElement, engine) {
...
 
             //offset.applyQuaternion( quat );
 
+            // get camera position in local space of target
+            offset.copy(this.camera.position).applyMatrix4( globeTarget.matrixWorldInverse );
+
             // angle from z-axis around y-axis
-            if(sphericalDelta.theta || sphericalDelta.phi )
+            if(sphericalDelta.theta || sphericalDelta.phi ) {
+
```
Comment (peppsac) [inline src/Renderer/ThreeExtented/GlobeControls.js]: Remove blank lines

File: src/Renderer/ThreeExtented/GlobeControls.js [hunk 15]
```diff
@@ -1179,9 +1144,13 @@ GlobeControls.prototype.setHeading = function(heading) {
     this.updateCameraTransformation();
 };
 
+GlobeControls.prototype.getCenter = function() {
```
Comment (peppsac) [inline src/Renderer/ThreeExtented/GlobeControls.js]: `getCenter`: center of what? center of view? on the globe?
Add a single line of comment or rename using a more descriptive  name

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 16]
```diff
@@ -137,9 +137,16 @@ ApiGlobe.prototype.moveLayerToIndex = function(layer, newId) {
...
+/**
+ * Removes a specific imagery layer from the current layer list. This removes layers inserted with addLayer().
+ *
+ * @param      {string}   id      The identifier
+ * @return     {boolean}  { description_of_the_return_value }
+ */
 ApiGlobe.prototype.removeImageryLayer = function(id) {
 
-    if (this.scene.getMap().removeColorLayer(id)) {
+    if (this.scene.getMap().layersConfiguration.removeColorLayer(id)) {
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:147]: This doesn't fix half of the bugs in `removeImageryLayer` sadly - but I can wait for another PR to fix this.

**Review by peppsac — CHANGES_REQUESTED** [review]

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 16 above)
Comment (gchoqueux) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:147]: Yes in is fixed in #200

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 7 above)
Comment (gchoqueux) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: I prefer keep the develop operation to understand : `projection = 2.0 * rayon * Math.sin(phi + alpha);`

File: src/Renderer/ThreeExtented/GlobeControls.js (hunk 12 above)
Comment (gchoqueux) [reply src/Renderer/ThreeExtented/GlobeControls.js:478]: For info on rework GlobeControl see #153 

> try clicking outside the globe

 Can you be more precise for this bug

File: src/Renderer/ThreeExtented/GlobeControls.js (hunk 13 above)
Comment (gchoqueux) [reply src/Renderer/ThreeExtented/GlobeControls.js]: This part must be reworked

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 9 above)
Comment (gchoqueux) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: ok

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 4 above)
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: the previous one was false calculation. I will return it later

File: src/Renderer/Camera.js (hunk 11 above)
Comment (gchoqueux) [reply src/Renderer/Camera.js:128]: the previous one was false calculation. I will return it later

**Comment by gchoqueux** [comment]
updated

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 17]
```diff
@@ -122,6 +121,12 @@ ApiGlobe.prototype.moveLayerDown = function(layer) {
     this.scene.renderScene3D();
 };
 
+/**
+ * { Moves a specific layer to a specific index in the layer list. This function has no effect if the layer is moved to its current index. }
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Could you remove the `{ }` ?

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 18]
```diff
@@ -122,6 +121,12 @@ ApiGlobe.prototype.moveLayerDown = function(layer) {
...
+ *
+ * @param      {string}  layer   The layer
+ * @param      {number}  newId   The new identifier
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Incorrect parameter names:
- `layer` →`layerId`
- `newId` →`newIndex` (and change the documentation accordingly)

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 16 above)
Comment (peppsac) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:147]: Ok 👍

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 19]
```diff
@@ -170,39 +182,34 @@ ApiGlobe.prototype.addElevationLayer = function(layer) {
...
-    //console.log(this.addImageryLayer().id);
-    var map = this.scene.getMap();
-    var manager = this.scene.managerCommand;
-    var providerWMTS = manager.getProvider(map.tiles).providerWMTS;
-    var layerWMTS = providerWMTS.layersData;
-    return layerWMTS[id].zoom.min;
+    id = id || 0;
+    var layer = this.getLayers()[id];
+    if(layer)
+        return layer.zoom.min;
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:185]: There are several problems with this function.
1) what is `id` ? `layerIndex` would be a better name (as far as I can tell)
2) what if `id` is invalid ?
3) what if this layer has no zoom property (a `wms` layer for instance ?)
4) no need to modify id, you could use `this.getLayers()[id || 0]`
5) given the function name, maybe you could return the min zoom value for all layers with a zoom value if id is undefined?

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 20]
```diff
@@ -170,39 +182,34 @@ ApiGlobe.prototype.addElevationLayer = function(layer) {
...
  */
 
-ApiGlobe.prototype.getLayers = function( /*param*/ ) {
-    var map = this.scene.getMap();
-    return map.layersConfiguration.getColorLayers();
+ApiGlobe.prototype.getMaxZoomLevel = function(id) {
+    id = id || 0;
+    var layer = this.getLayers()[id];
+    if(layer)
+        return layer.zoom.max;
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:207]: See comments for getMinZoomLevel

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 21]
```diff
@@ -170,39 +182,34 @@ ApiGlobe.prototype.addElevationLayer = function(layer) {
...
- * Gets the maximun zoom level of the chosen layer.
- * <iframe width="100%" height="400" src="//jsfiddle.net/iTownsIGN/y1xcqv4s/embedded/" allowfullscreen="allowfullscreen" frameborder="0"></iframe>
+ * Return the list of all layers in the scene in the order of how they are stacked on top of each other.
  * @constructor
  * @param {id} id - The id of the layer.
  */
 
-ApiGlobe.prototype.getMaxZoomLevel = function(id) {
-    //console.log(this.addImageryLayer().id);
+ApiGlobe.prototype.getLayers = function( /*param*/ ) {
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: `getColorLayers` ? `getImageryLayers` ?

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 22]
```diff
@@ -170,39 +182,34 @@ ApiGlobe.prototype.addElevationLayer = function(layer) {
...
+    if(layer)
+        return layer.zoom.max;
 };
-
 /**
- * Gets the maximun zoom level of the chosen layer.
- * <iframe width="100%" height="400" src="//jsfiddle.net/iTownsIGN/y1xcqv4s/embedded/" allowfullscreen="allowfullscreen" frameborder="0"></iframe>
+ * Return the list of all layers in the scene in the order of how they are stacked on top of each other.
  * @constructor
  * @param {id} id - The id of the layer.
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Remove invalid comments

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 23]
```diff
@@ -456,16 +463,17 @@ ApiGlobe.prototype.getHeading = function() {
  */
 
 ApiGlobe.prototype.getRange = function() {
+    return this.scene.currentControls().getRange();
+};
 
+ApiGlobe.prototype.getRangeFromEllispoid = function() {
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: You didn't correct the typo `getRangeFromEllipsoid`

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 24]
```diff
@@ -565,28 +573,150 @@ ApiGlobe.prototype.setCenterAdvanced = function(pPosition /*, pDisableAnimationo
...
  * <iframe width="100%" height="400" src="//jsfiddle.net/iTownsIGN/Lt3jL5pd/embedded/" allowfullscreen="allowfullscreen" frameborder="0"></iframe>
  * @constructor
  * @param {Number} pRange - The camera altitude.
  * @param {Boolean} [pDisableAnimation] - Used to force the non use of animation if its enable.
  */
-
-ApiGlobe.prototype.setRange = function(pRange /*, bool*/ ) {
+ApiGlobe.prototype.setRange = function(pRange /*, bool anim*/) {
     eventRange.oldRange = this.getRange();
+    loaded = false;
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:603]: Are you sure this is a good idea? I assumed that the `globe-loaded` event was supposed to be fired just once.
If that's not the case maybe documenting each event would be a good idea.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 25]
```diff
@@ -565,28 +573,150 @@ ApiGlobe.prototype.setCenterAdvanced = function(pPosition /*, pDisableAnimationo
...
     this.scene.currentControls().setRange(pRange);
-    this.viewerDiv.dispatchEvent(eventRange);
+    this.addEventListener('globe-loaded', updateTargetCamera(this));
+};
+
+/**
+ * { Displaces the central point to a specific amount of pixels from its current position.
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Remove `{  }`

File: src/Renderer/Camera.js (hunk 11 above)
Comment (gchoqueux) [reply src/Renderer/Camera.js:128]: I updated the branch with a new commit and a new comment in PR

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 26]
```diff
@@ -565,28 +573,150 @@ ApiGlobe.prototype.setCenterAdvanced = function(pPosition /*, pDisableAnimationo
...
+    // Screen pitch, in millimeters
+    pitch = (pitch || 0.28)/1000;
+
+    let camera = this.scene.currentCamera();
+    let center = this.scene.currentControls().getTargetCameraPosition();
+    let rayon = center.length();
+    let range = center.distanceTo(camera.camera3D.position);
+    // compute distance camera/globe's center
+    let distance = rayon + range;
+    // C is point in position camera
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: C = camera's position

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 27]
```diff
@@ -565,28 +573,150 @@ ApiGlobe.prototype.setCenterAdvanced = function(pPosition /*, pDisableAnimationo
...
+    // G is point in globe's center
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: G = globe's center

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 28]
```diff
@@ -565,28 +573,150 @@ ApiGlobe.prototype.setCenterAdvanced = function(pPosition /*, pDisableAnimationo
...
+    // S is point intersection with surface globe and camera (the furthest from line CG)
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: S = the furthest interesection[camera frustrum, globe surface]  from line CG

File: src/Renderer/Camera.js (hunk 11 above)
Comment (peppsac) [reply src/Renderer/Camera.js:128]: You ignored the 2nd part of my comment :)

> Documenting why (in commit message) would be interesting - and maybe change the commit's title btw: you're doing more than juste AddNew Api Functions.

File: src/Renderer/Camera.js [hunk 29]
```diff
@@ -116,19 +116,20 @@ Camera.prototype.resize = function(width, height) {
...
     var altiW = node.bbox.top() === 10000 ? 0. : node.bbox.bottom() / 10000.;
     var dotProductW = Math.min(altiW + Math.abs(this.camera3D.getWorldDirection().dot(node.centerSphere.clone().normalize())), 1.);
     if (this.camera3D.position.length() > 6463300) dotProductW = 1;
     var SSE = Math.sqrt(dotProductW) * this.preSSE * (node.geometricError / distance);
-    //var SSE = this.preSSE * (node.geometricError / distance);
+    */
+    var SSE = this.preSSE * (node.geometricError / distance);
```
Comment (peppsac) [inline src/Renderer/Camera.js:137]: Could you add a comment here:

```
// TODO: node.geometricError is computed using a hardcoded 18 level
```

File: src/Scene/Quadtree.js [hunk 30]
```diff
@@ -28,8 +29,7 @@ function commandQueuePriorityFunction(cmd) {
         if (!node.loaded) {
             return 1000;
         } else {
-            // TODO: this magic value comes from NodeProcess
-            if (6.0 < node.sse) {
+            if (SSE_SUBDIVISION_THRESHOLD < node.sse) {
```
Comment (peppsac) [inline src/Scene/Quadtree.js:32]: 👍

**Review by peppsac — CHANGES_REQUESTED** [review]

File: src/Renderer/Camera.js (hunk 11 above)
Comment (peppsac) [reply src/Renderer/Camera.js:128]: Oops, I forgot to read the commit messages.. nevermind :-)

Comment (gchoqueux) [reply src/Renderer/Camera.js:128]: and I add  link  in [ main comment PR #194](https://github.com/iTowns/itowns2/pull/194#issue-181982034)

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 24 above)
Comment (gchoqueux) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:603]: Yes, the event represents the tiles are loaded. In this case, I need to know when the tiles are loaded to compute the new target camera.

**Review by peppsac — APPROVED** [review]

**Comment by peppsac** [comment]
Looking better, thanks. Merging

**Pull request merged**